test/autoproj/cli/test_daemon.rb

```diff
@@ -33,69 +37,442 @@ def define_package(name, vcs)
+                            branch: 'master',
+                            pull_request: @pull_request,
+                            created_at: Time.now
+                        )
+                    end
+                    it 'is ignored if it is to the buildconf' do
+                        event = create_pull_request_event(
+                            base_owner: 'rock-core',
+                            base_name: 'buildconf',
+                            base_branch: 'master',
+                            state: 'closed',
+                            created_at: Time.now
+                        )
+
+                        flexmock(cli).should_receive('client.delete_branch_by_name')
+                                     .with(any).never
+                        cli.handle_pull_request_event(event)
+                    end
+                    it 'handles errors if buildconf branch does not exist' do
+                        flexmock(cli).should_receive('client.delete_branch_by_name')
+                                     .with('rock-core', 'buildconf',
+                                           'autoproj/rock-core/'\
+                                           'drivers-iodrivers_base/pulls/1')
+                                     .and_raise(Octokit::UnprocessableEntity)
+
+                        cli.handle_pull_request_event(@pull_request_event)
+                    end
+                    it 'deletes branch and removes PR from cache' do
+                        cli.cache.add(@pull_request, [])
+
+                        flexmock(cli).should_receive('client.delete_branch_by_name')
+                                     .with('rock-core', 'buildconf',
+                                           'autoproj/rock-core/drivers-iodrivers_base/'\
+                                           'pulls/1')
+
+                        refute cli.cache.changed?(@pull_request, [])
+                        cli.handle_pull_request_event(@pull_request_event)
+                        assert cli.cache.changed?(@pull_request, [])
+                    end
                 end
             end
+            # rubocop: enable Metrics/BlockLength
 
             describe '#setup_hooks' do
-                it 'updates the workspace on push event' do
+                it 'sets up push and pull request handlers' do
                     watcher = flexmock
                     flexmock(cli).should_receive(:watcher).and_return(watcher)
-                    flexmock(Process)
-                        .should_receive(:exec)
-                        .with(Gem.ruby, $PROGRAM_NAME, 'daemon',
-                              'start', '--update').once
 
                     watcher.should_receive(:add_push_hook)
-                           .and_yield('rock-core', 'autoproj')
+                           .and_yield('push_event', options: 'options')
+                    watcher.should_receive(:add_pull_request_hook)
+                           .and_yield('pull_request_event')
```

Comment:
This is not incidental, this is actually the very thing I want to test here. The only purpose of `setup_hooks` is to call `add_push_hook` and `add_pull_request_hook` passing a block that forwards the arguments to the proper handlers.. This *is* its role. I could, rather than yielding, store the block in the test case and "force" call it to check that the expectations are met but I don't see any difference in that.

`call_push_hooks` is a method of `GithubWatcher` which is not the test subject, calling that should be done in `GithubWatcher`'s tests (and it is).

Reply:
Well. `add_push_hook` and `add_pull_request_hook` _in real life_ do not yield. Making them yield here is at the very least very confusing (this discussion being a consequence of that). You're basically mis-using `and_yield` to verify what block is being passed as argument to `add_push_hook`.

My understanding is that you are trying to be very unit-testy and not touch anything but "your subject". I don't subscribe to this strategy, but let's keep on with it regardless.

There is no easy way to test the block passed to a method. If you do control the block, you can. But you're not controlling it in this case ... so `and_yield` is indeed a good solution - just put a lot of commenting to explain why `and_yield` is being used here on a method that never yields.

Now, I believe that doing it through `call_push_event` would be even better, as it tests that `setup_hooks` does _whatever is necessary_ to be called when the watcher has an event. It is  less strictly "unit-testy", but IMO performs a better test as it assumes less about the underlying APIs and more about the actual functionality / behavior that is required from it.

Reply:
> you are trying to be very unit-testy

We've been in this position before and we really do have different opinions on this. There's definitely an argument to both sides. I honestly believe that by testing a lot of stuff in a test case  hurts maintainability of the tests. It also usually introduces dependencies on implementation details (specially if the resources requires mocks).

So, if you don't mind I will just add the comment you suggested...

Reply:
> So, if you don't mind I will just add the comment you suggested...

I don't, this is why I suggested them. And I agree that we have to continue disagreeing on this.